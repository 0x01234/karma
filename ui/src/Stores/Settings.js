import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
