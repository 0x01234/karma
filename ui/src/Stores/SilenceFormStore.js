import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
