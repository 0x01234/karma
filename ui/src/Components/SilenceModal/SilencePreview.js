import React, { Component } from "react";
import PropTypes from "prop-types";

import { observer } from "mobx-react";

import JSONPretty from "react-json-pretty";
import "react-json-pretty/src/JSONPretty.monikai.css";

import { SilenceFormStore } from "Stores/SilenceFormStore";

const SilencePreview = observer(
  class SilencePreview extends Component {
    static propTypes = {
      silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired
    };

    render() {
      const { silenceFormStore } = this.props;

      return (
        <div className="mt-3">
          <JSONPretty json={silenceFormStore.data.toAlertmanagerPayload} />
        </div>
      );
    }
  }
);

export { SilencePreview };
