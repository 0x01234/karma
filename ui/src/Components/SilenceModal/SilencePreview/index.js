import React, { Component } from "react";
import PropTypes from "prop-types";

import { observable, action } from "mobx";
import { observer } from "mobx-react";

import hash from "object-hash";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowLeft } from "@fortawesome/free-solid-svg-icons/faArrowLeft";
import { faCheckCircle } from "@fortawesome/free-solid-svg-icons/faCheckCircle";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons/faExclamationCircle";

import { AlertStore, FormatBackendURI, FormatAlertsQ } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { StaticLabel } from "Components/Labels/StaticLabel";
import { MatcherToFilter, AlertManagersToFilter } from "../Matchers";

const FetchError = ({ message }) => (
  <div className="text-center">
    <h2 className="display-2 text-danger">
      <FontAwesomeIcon icon={faExclamationCircle} />
    </h2>
    <p className="lead text-muted">{message}</p>
  </div>
);
FetchError.propTypes = {
  message: PropTypes.node.isRequired
};

const Preview = ({ alertStore, labelsList }) => (
  <ul className="list-group list-group-flush px-2">
    {labelsList.map(labels => (
      <li key={hash(labels)} className="list-group-item px-2 pt-2 pb-1">
        {Object.entries(labels).map(([name, value]) => (
          <StaticLabel
            key={name}
            alertStore={alertStore}
            name={name}
            value={value}
          />
        ))}
      </li>
    ))}
  </ul>
);
Preview.propTypes = {
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  labelsList: PropTypes.arrayOf(PropTypes.object).isRequired
};

const SilencePreview = observer(
  class SilencePreview extends Component {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired
    };

    matchedAlerts = observable(
      {
        alertLabels: [],
        error: null,
        fetch: null,
        groupsToUniqueLabels(groups) {
          const alerts = {};
          for (const group of groups) {
            for (const alert of group.alerts) {
              const alertLabels = Object.assign(
                {},
                group.labels,
                group.shared.labels,
                alert.labels
              );
              const alertHash = hash(alertLabels);
              alerts[alertHash] = alertLabels;
            }
          }
          this.alertLabels = Object.values(alerts);
        },
        setError(value) {
          this.error = value;
        }
      },
      {
        groupsToUniqueLabels: action,
        setError: action
      }
    );

    onFetch = () => {
      const { silenceFormStore } = this.props;

      const filters = [
        ...silenceFormStore.data.matchers.map(m => MatcherToFilter(m)),
        AlertManagersToFilter(silenceFormStore.data.alertmanagers)
      ];

      const alertsURI =
        FormatBackendURI("alerts.json?") + FormatAlertsQ(filters);

      this.matchedAlerts.fetch = fetch(alertsURI, { credentials: "include" })
        .then(result => {
          return result.json();
        })
        .then(result => {
          this.matchedAlerts.groupsToUniqueLabels(Object.values(result.groups));
          this.matchedAlerts.setError(null);
        })
        .catch(err => {
          console.trace(err);
          return this.matchedAlerts.setError(
            `Request failed with: ${err.message}`
          );
        });
    };

    componentDidMount() {
      this.onFetch();
    }

    render() {
      const { alertStore, silenceFormStore } = this.props;

      return (
        <React.Fragment>
          <div className="mb-2">
            {this.matchedAlerts.error !== null ? (
              <FetchError message={this.matchedAlerts.error} />
            ) : (
              <React.Fragment>
                <p className="lead text-center">
                  This silence will match{" "}
                  <strong>{this.matchedAlerts.alertLabels.length}</strong> alert
                  {this.matchedAlerts.alertLabels.length > 1 ? "s" : ""}
                </p>
                <div>
                  <Preview
                    alertStore={alertStore}
                    labelsList={this.matchedAlerts.alertLabels}
                  />
                </div>
              </React.Fragment>
            )}
          </div>
          <div className="d-flex flex-row-reverse">
            <button
              type="button"
              className="btn btn-outline-primary"
              onClick={silenceFormStore.data.setStageSubmit}
            >
              <FontAwesomeIcon icon={faCheckCircle} className="pr-1" />
              Submit
            </button>
            <button
              type="button"
              className="btn btn-outline-secondary mr-2"
              onClick={silenceFormStore.data.resetProgress}
            >
              <FontAwesomeIcon icon={faArrowLeft} className="pr-1" />
              Back
            </button>
          </div>
        </React.Fragment>
      );
    }
  }
);

export { SilencePreview };
