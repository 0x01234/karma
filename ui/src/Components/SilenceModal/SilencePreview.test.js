import React from "react";

import { render } from "enzyme";

import moment from "moment";

import { SilenceFormStore } from "Stores/SilenceFormStore";
import { SilencePreview } from "./SilencePreview";

describe("<SilencePreview />", () => {
  it("matches snapshot", () => {
    const silenceFormStore = new SilenceFormStore();
    silenceFormStore.data.startsAt = moment([2000, 1, 1, 0, 0, 0]);
    silenceFormStore.data.endsAt = moment([2000, 1, 1, 1, 0, 0]);
    silenceFormStore.data.createdBy = "me@example.com";
    silenceFormStore.data.comment = "SilencePreview test";

    const tree = render(<SilencePreview silenceFormStore={silenceFormStore} />);
    expect(tree).toMatchSnapshot();
  });
});
