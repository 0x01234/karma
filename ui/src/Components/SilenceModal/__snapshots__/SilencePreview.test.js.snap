// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SilencePreview /> matches snapshot 1`] = `
<div
  class="mt-3"
>
  <pre
    class="json-pretty"
  >
    {
      "
    <span
      class="json-key"
    >
      matchers
    </span>
    ": [],
      "
    <span
      class="json-key"
    >
      startsAt
    </span>
    ": 
    <span
      class="json-string"
    >
      "2000-02-01T00:00:00.000Z"
    </span>
    ,
      "
    <span
      class="json-key"
    >
      endsAt
    </span>
    ": 
    <span
      class="json-string"
    >
      "2000-02-01T01:00:00.000Z"
    </span>
    ,
      "
    <span
      class="json-key"
    >
      createdBy
    </span>
    ": 
    <span
      class="json-string"
    >
      ""
    </span>
    ,
      "
    <span
      class="json-key"
    >
      comment
    </span>
    ": 
    <span
      class="json-string"
    >
      "SilencePreview test"
    </span>
    
    }
  </pre>
</div>
`;
